### PR TITLE
Small optimisation of the graph traversal for freeze 

### PR DESCRIPTION
**Pull request by icristescu**
When traversing the closure graph of a commit during a freeze we don't need to remember the visited objects: those are already copied in their destination layer, and the skip function will skip them anyway. It is only a small optimisation (it saves ~4min out of 30min for the first freeze in bench_layers) and it saves a bit of memory allocations as well. 
This PR:
no_hashtbl
vs master:
with_hashtbl

### CODE REVIEW

File: src/irmin-pack/irmin_pack_layers.ml [hunk 1]
```diff
@@ -321,13 +321,13 @@ struct
                     let commit : 'a Commit.t = (node, commit) in
                     f contents node commit)))
 
-      let log_current_upper t = if t.flip then "upper1" else "upper0"
+      let log_next_upper t = if t.flip then "upper0" else "upper1"
```
Comment (samoht) [inline src/irmin-pack/irmin_pack_layers.ml:324]: that seems unrelated -- could you make a separate PR for this?

File: src/irmin/object_graph.ml [hunk 2]
```diff
@@ -182,6 +191,41 @@ struct
     in
     visit ()
 
+  let iter_rev ~pred ~min ~max ~node ~skip () =
+    Log.debug (fun f -> f "iter on closure (%d elements)" (List.length max));
```
Comment (samoht) [inline src/irmin/object_graph.ml:195]: I wondering if this should not be the default implementation for iter (and then users can use `skip` to store an hashtable). Need to run benchmarks to understand the exact impact. 

Note that this start to look very similar to the API proposed by @liautaud for the GC.

Comment (icristescu) [reply src/irmin/object_graph.ml]: The default iter now does not use the hash table and added an explicit [iter with tracking](https://github.com/mirage/irmin/pull/1122/files#diff-e838a79261906c2af6541f96fc7772f019e640f6e12820a8da7195cfc78e0256R321).

**Comment by CraigFe** [comment]
Rebased + resolved conflicts, but otherwise no changes made.

**Comment by samoht** [comment]
Rebased again but a test is failing:

```
> [FAIL]        LAYERED_PACK            16   Branches with squash.
...
FAIL copy v2 to dst

   Expected: `{"some":"X2"}'
   Received: `null'
```

**Comment by icristescu** [comment]
Another test fails, but I can't reproduce it on my computer. 
`2020-12-09 20:02.22: Failed: LAYERED_PACK Test commits and graphs.`

**Comment by samoht** [comment]
Are you ok to use #1188 instead of this PR? I think they share the same goal (removing a global hashtable for Repo.iter) but I believe the API is a bit simpler in #1188.

**Comment by icristescu** [comment]
> Are you ok to use #1188 instead of this PR? I think they share the same goal (removing a global hashtable for Repo.iter) but I believe the API is a bit simpler in #1188.

Sure. The API here is more flexible (i.e. you can add any tracking mechanism you want thanks to `skip`, it's not hardcoded in the `iter` function) but we don't need this flexibility right now, so let's go with #1188.

**Comment by samoht** [comment]
(happy to resurect this one later on if we need that extra flexibilty)
